src/zarr/abc/store.py

```diff
@@ -1,31 +1,44 @@
 from abc import ABC, abstractmethod
 from collections.abc import AsyncGenerator
-from typing import Protocol, runtime_checkable
+from typing import NamedTuple, Protocol, runtime_checkable
+
+from typing_extensions import Self
 
 from zarr.buffer import Buffer, BufferPrototype
-from zarr.common import BytesLike, OpenMode
+from zarr.common import BytesLike, OpenModeLiteral
+
+
+class OpenMode(NamedTuple):
+    overwrite: bool
+    can_create: bool
+    can_open_existing: bool
+    is_writable: bool
+
+    @classmethod
+    def from_str(cls, mode: str) -> Self:
+        if mode not in ("r", "r+", "a", "w", "w-"):
+            raise ValueError("mode must be one of 'r', 'r+', 'w', 'w-', 'a'")
+        return cls(
+            overwrite=mode == "w",
+            can_create=mode in ("a", "w", "w-"),
+            can_open_existing=mode in ("r", "r+", "a"),
+            is_writable=mode in ("r+", "a", "w", "w-"),
+        )
 
```

Comment:
what do you think? @d-v-b 
(I think the name `write_mode` does not describe it fully, as it also describes the behavior at opening an array, so I called it `OpenMode`, but I am open to discussion)

Reply:
re: naming, could `AccessMode` also work? But `OpenMode` is also fine with me.

regarding the fields of `OpenMode`:
-  if `overwrite` is `True`, then `can_create` and `is_writable` must also be `True`, and `can_open_existing` must be `False`. This means that `overwrite` is degenerate with respect to the other three attributes, and thus `overwrite` can be removed as a field (perhaps replaced by `@property`-decorated method, if we use a dataclass instead of a named tuple here).
- I would shorten `can_create` to just `create`, `is_writeable` to `write`, and `can_open_existing` to `allow_existing` / `allow_extant`. This is personal preference.


Here's another idea, for inspiration; feel free to ignore it :) 

```python

class AccessMode:
  mutable: bool

@dataclass
class ReadMode(AccessMode):
    mutable: False

@dataclass
class WriteMode(AccessMode):
  mutable: True
  update: bool # instead of can_open_existing
  create: bool # instead of can_create


class Store:
  access_mode: ReadMode | WriteMode
  ...

```
Code consuming this API would look like this:

```python

if (m := store.access_mode).mutable:
    if m.update:
        ...
    if m.create:
       ...
```

Reply:
I like this. also `AccessMode` feels better than `OpenMode`. 

The overwrite attribute is necessary to distinguish between `w` (overwrite) and `w-` (raise error if file exist)